test/performance/observed_concurency_test.go

```diff
@@ -51,16 +56,84 @@ func createTestCase(val float32, name string) testgrid.TestCase {
 	return tc
 }
 
-func waitForServiceLatestCreatedRevision(clients *test.Clients, names test.ResourceNames) (string, error) {
-	var revisionName string
-	err := test.WaitForServiceState(clients.ServingClient, names.Service, func(s *v1alpha1.Service) (bool, error) {
-		if s.Status.LatestCreatedRevisionName != names.Revision {
-			revisionName = s.Status.LatestCreatedRevisionName
-			return true, nil
+// generateTraffic loads the given endpoint with the given concurrency for the given duration.
+// All responses are forwarded to a channel, if given.
+func generateTraffic(client *spoof.SpoofingClient, url string, concurrency int, duration time.Duration, resChannel chan *spoof.Response) (int32, error) {
+	var (
+		group    errgroup.Group
+		requests int32
+	)
+
+	for i := 0; i < concurrency; i++ {
+		group.Go(func() error {
+			done := time.After(duration)
+			req, err := http.NewRequest(http.MethodGet, url, nil)
+			if err != nil {
+				return fmt.Errorf("error creating http request: %v", err)
+			}
+			for {
+				select {
+				case <-done:
+					return nil
+				default:
+					res, _ := client.Do(req)
+					if resChannel != nil {
+						resChannel <- res
+					}
+					atomic.AddInt32(&requests, 1)
+				}
+			}
+		})
+	}
+
+	err := group.Wait()
+	requestsMade := atomic.LoadInt32(&requests)
+
+	if err != nil {
+		return requestsMade, fmt.Errorf("error making requests for scale up: %v", err)
+	}
+	return requestsMade, nil
+}
+
+// event represents the start or end of a request
+type event struct {
+	concurrencyModifier int32
+	timestamp           time.Time
+}
+
+// parseResponse parses a string of the form TimeInNano,TimeInNano into the respective
+// start and end event
+func parseResponse(body string) (*event, *event, error) {
+	body = strings.TrimSpace(body)
+	parts := strings.Split(body, ",")
+	start, err := strconv.ParseInt(parts[0], 10, 64)
```

Comment:
what are 10 and 64 here? I am guessing the start time start and end indexes. Can we make them const?

Reply:
10 is the base of the numeric system, 64 are the bits to use for the parsed number.

Essentially this is saying: Parse a decimal number of 64bit width.